rfcs/0064-autotvm-tune-with-subgraph-granularity.md

```diff
@@ -0,0 +1,135 @@
+2.5 Create `workload` for subgraph tuning task with `task_name` and `iotensors`.
+2.6 Lower schedule with the best config queried by `workload`.
+2.7 Codegen.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In steps 1.1 and 2.2 mentioned in the previous section, returning the best implementation name is kind of tricky. We can get the `best_plevel_impl` in function `select_implementation`, but sometimes the actual implementation name is not `best_plevel_impl.name`. For example, implementation for `conv2d_nchw.x86` is added like this.
+
+```python
+@conv2d_NCHWc_strategy.register("cpu")
+def conv2d_NCHWc_strategy_cpu(attrs, inputs, out_type, target):
+    """conv2d_NCHWc x86 strategy"""
+    strategy = _op.OpStrategy()
+    strategy.add_implementation(
+        wrap_compute_conv2d(topi.x86.conv2d_nchw),
+        wrap_topi_schedule(topi.x86.schedule_conv2d_nchw),
+        name="conv2d_nchw.x86",
+    )
+    return strategy
+```
+
+But `topi.x86.conv2d_nchw` wrappers another implementation.
+
+```python
+def conv2d_nchw(data, kernel, strides, padding, dilation, out_dtype):
+    layout = "NCHW"
+    packed_out = conv2d_NCHWc(data, kernel, strides, padding, dilation, layout, layout, out_dtype)
+    return unpack_NCHWc_to_nchw(packed_out, out_dtype)
+```
+
+The actual implementation name is `conv2d_NCHWc.x86`. So in function `select_implementation` we fixed this problem by getting the actual implementation name from the workload which is created in function `register_topi_compute`.
+
+```python
+def select_implementation():
+  # ignore some codes
+
+  if GLOBAL_SCOPE.tune_subgraph:
+    # In some cases, one strategy's compute may call another compute.
+    # So the impl name need to match with actual compute.
+    if workloads[best_plevel_impl]:
+        workload = workloads[best_plevel_impl]
+        if best_plevel_impl.name != "injective.cpu" and best_plevel_impl.name != workload[0]:
+          best_plevel_impl.name = workload[0]
+  # value changed in python side will not effect C++ side,
+  # so here need to pass new name to C++
+  return best_plevel_impl, outputs[best_plevel_impl], best_plevel_impl.name
+```
+
+Because the value changed on the python side will not affect the C++ side, we need to pass the new name to C++. This results in the function `select_implementation` number of returned values changing to three. This should be noticed.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are three cases in which tuning with subgraph may not get better performance.
+
+Case 1: Anchor operator has more than one implementation.
+We register the subgraph tuning task by `outputs` in step 1.4. No matter how many implementations the anchor operator has, step 1.1 will only pick the implementation with the highest level and return outputs computed by it. So the subgraph tuning tasks may not contain the potential best implementation.
```

Comment:
This is my major concern too, as one important feature in AutoTVM is transparent the implementation selection to the tuning results. This is also a challenge issue that auto-scheduler hasn't figured out yet. As you pointed out, selecting an implementation based on the plevel before tuning could result in sub-optimal, is there anyway to still generate N subgraph tasks (N=#impl) for the anchor op?

Reply:
Generate N subgraph tasks means we need N subgraph outputs and that means we need to run `LowerToTECompute::Lower` N times. It will be a big change to the current framework but I think it will work. Even we can resolve this case, case 2 is still a problem preventing anchor op to find the best config from N tuning results.